### ISSUE TITLE
Multiple openssl build issues surrounding all versions of python, build from source, freebsd

### ISSUE THREAD

**Issue by paulkudla** [pyca/cryptography#5460]
Please bear with me, hopefully i am explaining this correctly

A bit of background 👍 

I am a programmer by trade going all the way back to the Z80 assembler days 34 years ago at radio shack, as a programmer i do understand that you cant support all flavors of all programming introduced in today's environment.

However I spent 2 days try to simply generate csr certs in python (i am writing this into django projects)

I am running python 2.7,3.7,3.8 (freebsd kinda makes you do that with the ports)

the issue is that freebsd ports collection will support 2.7 on one server (but not really - see below) and 3.7 on another server but not 3.8 with no way to build the python modules via the ports collection that some how picks up on the openssl supported protocols?

Apparently cryptography does not look at the openssl installed on the system to decide what can be supported from there?

In the mean time (and i have researched this a lot i think) i cant find instructions on how to build cryptography manually from source disabling the ssl versions that are not needed or no longer supported (TLS 1.2 seems to be the latest?)

for example i get that SSLv3 is no longer used or supported however when i build the cryptography python libs from scratch it is still included? even though openssl was built wihout it.

example issues when trying to run cryptography or anything that uses it.

/usr/local/lib/python2.7/site-packages/cryptography/hazmat/bindings/_openssl.so: Undefined symbol "EVP_MD_CTX_new"

SSLv3 no longer supported / missing from openssllibs

ImportError: /usr/local/lib/python3.8/site-packages/cryptography/hazmat/bindings/_openssl.abi3.so: Undefined symbol "SSLv3_client_method"

So long story short can anyone point me in the direction on how to modify the config file or instructions to setup.py to be able to disable support for various protocols when manually building cryptography?

thanks in advance, send me how to contribute and i will, nothing is for free !

**Comment by reaperhulk** [comment]
OpenSSL has a large number of compile flags that result in an absurd number of possible permutations. `cryptography` supports a reasonable subset, and it does determine much of this at compile time. (Due to limitations in cffi we can't easily probe for support though and need to rely exclusively on headers) Ultimately, you need to know quite a bit about OpenSSL and how C toolchains work to safely make changes.

For example, you seem concerned about how this interacts with ports. `cryptography` knows nothing about ports. Instead, it cares about where it can find the object files and headers. The compiler and linker likely have default paths that include the ports tree, but you can specify arbitrary paths yourself using `CFLAGS` and `LDFLAGS`.

Your other question was around SSLv3 support. OpenSSL has two separate flags to fully disable SSLv3. If you are compiling OpenSSL yourself you need to pass both `no-ssl3` and `no-ssl3-method`.  It is possible that the ports-compiled OpenSSL isn't setting `OPENSSL_NO_SSL3_METHOD` in `opensslconf.h`, but is removing those symbols anyway. If so, this is not a supported configuration because it's not how OpenSSL upstream works.

If you run into a genuine bug we're happy to review reports with easily reproducible steps, but as you noted we can't support entirely arbitrary OpenSSL configurations. In case you were wondering, our current test infrastructure runs on the following:
- OpenSSL 1.1.1-latest with the [following flags](https://github.com/pyca/infra/blob/7b698bfff62118ed4457df2daee17ca08798b344/cryptography-manylinux/openssl-version.sh#L5)
- OpenSSL 1.1.1g (default config)
- OpenSSL 1.1.1g (no-engine no-rc2 no-srtp no-ct)
- OpenSSL 1.1.0l (default config)
- OpenSSL 1.0.2u (default config)
- OpenSSL 1.1.1 (RHEL8-latest)
- OpenSSL 1.1.1 (FIPS mode RHEL8-latest)
- OpenSSL 1.0.2 (RHEL7-latest)
- OpenSSL 1.1.1f (Ubuntu Focal)
- OpenSSL 1.1.1 (Ubuntu Bionic)
- OpenSSL 1.1.1g (Fedora 32)
- OpenSSL 1.1.1g (Debian Bullseye)
- OpenSSL 1.1.1d (Debian Buster)
- OpenSSL 1.1.0l (Debian Stretch)
- LibreSSL 2.9.2
- LibreSSL 3.0.2
- LibreSSL 3.1.4
- LibreSSL 3.2.0
